Sources/Build/BuildPlan.swift

```diff
@@ -1833,53 +1850,64 @@ public class BuildPlan {
-    private func xcFrameworkLibrary(for target: BinaryTarget) -> LibraryInfo? {
-        func calculateLibraryInfo() -> LibraryInfo? {
-            // Parse the XCFramework's Info.plist.
-            let infoPath = target.artifactPath.appending(component: "Info.plist")
-            guard let info = XCFrameworkInfo(path: infoPath, diagnostics: diagnostics, fileSystem: fileSystem) else {
-                return nil
-            }
+    /// Extracts the library information from an XCFramework.
+    private func parseXCFramework(for target: BinaryTarget) throws -> LibraryInfo {
+        try self.externalLibrariesCache.memoize(key: target) {
+            let metadata = try XCFrameworkMetadata.parse(fileSystem: self.fileSystem, rootPath: target.artifactPath)
 
             // Check that it supports the target platform and architecture.
-            guard let library = info.libraries.first(where: {
-                return $0.platform == buildParameters.triple.os.asXCFrameworkPlatformString && $0.architectures.contains(buildParameters.triple.arch.rawValue)
+            guard let library = metadata.libraries.first(where: {
+                $0.platform == buildParameters.triple.os.asXCFrameworkPlatformString && $0.architectures.contains(buildParameters.triple.arch.rawValue)
             }) else {
-                diagnostics.emit(error: """
+                throw StringError("""
                     artifact '\(target.name)' does not support the target platform and architecture \
                     ('\(buildParameters.triple)')
                     """)
-                return nil
             }
 
             let libraryDirectory = target.artifactPath.appending(component: library.libraryIdentifier)
             let binaryPath = libraryDirectory.appending(component: library.libraryPath)
             let headersPath = library.headersPath.map({ libraryDirectory.appending(component: $0) })
+
             return LibraryInfo(binaryPath: binaryPath, headersPath: headersPath)
         }
+    }
 
-        // If we don't have the library information yet, calculate it.
-        if let xcFramework = xcFrameworkCache[target] {
-            return xcFramework
-        }
+    /// Extracts the executables info from an executablesArchive
+    private func parseToolsArchive(for target: BinaryTarget) throws -> [ToolInfo] {
+        try self.externalToolsCache.memoize(key: target) {
+            let metadata = try ToolsArchiveMetadata.parse(fileSystem: self.fileSystem, rootPath: target.artifactPath)
 
-        let xcFramework = calculateLibraryInfo()
-        xcFrameworkCache[target] = xcFramework
-        return xcFramework
+            // filter the tools that are relevant to the triple
+            let supportedTools = metadata.tools.filter { $0.value.contains(where: { $0.supportedTriplets.contains(buildParameters.triple) }) }
+            // flatten the tools for each access
+            return supportedTools.reduce(into: [ToolInfo](), { partial, entry in
```

Comment:
It looks as if this restricts the set of tools to just the executables.  Would an extension access ancillary files by looking in the path that contains the tool?  (for example because they arrange their archive that way)?  Or should there also be a support-files path in addition to the tools?

Reply:
this is a good question. I think that at the min the metadata should provide the exact path to the binary so its easy to run. we could add additional metadata field to track ancillary files location or have the extension walk up the directory structure. can go either way

Reply:
I agree about needing the exact path for the tool, especially since things like a `.exe` suffix would be used on Windows but most other platforms wouldn't have one.  Since the extension controls (or is at least aware of) the structure of the payload in each variant of the binary archive, and in the interest of starting small, we can leave any additional path as a future enhancement and just have the extension form a path relative to the tool for the first release.  We wouldn't break anyone by adding more flexibility later.